File: apps/vnext-web/lib/file-tree.ts
/**
 * A FLAT LIST OF PATHS, TURNED INTO SOMETHING WORTH LOOKING AT.
 *
 * The engine answers with paths (`WorkspaceListing`) because that is the
 * smallest true thing and because search wants a list, not a graph. Every
 * decision about how those paths GROUP, SORT, COLLAPSE and FILTER is a
 * presentation decision, so all of it is here — pure, and testable without a
 * repository or a browser.
 *
 * THREE OF THOSE DECISIONS ARE BORROWED, deliberately, from t3 code's file
 * browser, because they are what makes a monorepo tree usable rather than merely
 * correct:
 *
 *   - SINGLE-CHILD CHAINS COLLAPSE. `packages/core/src/loom/steps/` is six clicks
 *     and six rows to reach one directory that has anything in it. Merged into
 *     one row it is one click, and no information is lost — the full path is
 *     right there in the label.
 *   - ONE LEVEL IS OPEN. Enough to see the shape of the repository, few enough
 *     rows that the panel is not a wall of text before you have asked anything.
 *   - SEARCH HIDES NON-MATCHES rather than highlighting them. In a thousand-file
 *     tree, highlighting is asking somebody to scroll for a colour.
 *
 * WHAT IS NOT BORROWED: t3 code's tree is a third-party widget rendering into a
 * shadow root, and the panel has to inject CSS variables through `unsafeCSS` to
 * make it match the app it lives in. Ours renders with the same primitives as
 * every other row in this panel, so it matches by construction.
 */

export type FileTreeNode =
  | { kind: "file"; path: string; name: string }
  | { kind: "directory"; path: string; name: string; children: FileTreeNode[] };

/**
 * Directories before files, then natural order.
 *
 * `numeric` so `step-2` sorts before `step-10`, and `sensitivity: "base"` so a
 * capitalised name does not sort into its own block at the top — `README.md`
 * belongs next to `package.json`, not above every lowercase file in the
 * directory, which is what a plain codepoint sort does.
 */
function compareNodes(left: FileTreeNode, right: FileTreeNode): number {
  if (left.kind !== right.kind) return left.kind === "directory" ? -1 : 1;
  return left.name.localeCompare(right.name, undefined, { numeric: true, sensitivity: "base" });
}

type Building = { dirs: Map<string, Building>; files: string[] };

function emptyBuilding(): Building {
  return { dirs: new Map(), files: [] };
}

/**
 * Merge a directory that contains exactly one directory and nothing else.
 *
 * Recursive, and the LABEL is what grows — `core/src/loom` — while the `path`
 * stays the real path of the directory the row actually represents. Expanding it
 * reveals the contents of the deepest one, which is the only thing anybody
 * wanted from those five clicks.
 */
function collapse(node: FileTreeNode): FileTreeNode {
  if (node.kind === "file") return node;
  const children = node.children.map(collapse);
  const only = children[0];
  if (children.length === 1 && only?.kind === "directory") {
    return { kind: "directory", path: only.path, name: `${node.name}/${only.name}`, children: only.children };
  }
  return { kind: "directory", path: node.path, name: node.name, children };
}

function toNodes(building: Building, prefix: string): FileTreeNode[] {
  const nodes: FileTreeNode[] = [];
  for (const [name, child] of building.dirs) {
    const path = prefix ? `${prefix}/${name}` : name;
    nodes.push({ kind: "directory", path, name, children: toNodes(child, path) });
  }
  for (const name of building.files) nodes.push({ kind: "file", path: prefix ? `${prefix}/${name}` : name, name });
  return nodes.sort(compareNodes);
}

/**
 * Build the tree. Returns the ROOT'S CHILDREN, not a root node: the root is the
 * workspace itself, it is already named in the panel's header, and a row saying
 * `/Users/you/Projects/thing` above everything would be one click of overhead on
 * every visit.
 */
export function buildFileTree(paths: readonly string[]): FileTreeNode[] {
  const root = emptyBuilding();
  for (const path of paths) {
    const segments = path.split("/").filter(Boolean);
    const name = segments.pop();
    if (name === undefined) continue;
    let cursor = root;
    for (const segment of segments) {
      let next = cursor.dirs.get(segment);
      if (!next) {
        next = emptyBuilding();
        cursor.dirs.set(segment, next);
      }
      cursor = next;
    }
    cursor.files.push(name);
  }
  return toNodes(root, "").map(collapse);
}

/**
 * How many matches a search will draw.
 *
 * Typing one letter matches most of a repository, and building a thousand-node
 * tree on every keystroke to render a scroller nobody will reach the bottom of is
 * work for nothing. The cap is reported so the surface can say what it dropped —
 * a search that silently stops at 400 reads as "there are only 400".
 */
export const MAX_SEARCH_MATCHES = 400;

/**
 * The paths a query keeps.
 *
 * MATCHED ON THE WHOLE PATH, not the filename, which is what makes `engine`
 * bring back everything under `apps/engine` and `panel` bring back
 * `components/right-panel.tsx`. Case-insensitive, plain substring: a person
 * typing into a file search is remembering a fragment, not writing a pattern, and
 * a regex here would turn a typed `(` into an error state.
 */
export function matchFiles(
  paths: readonly string[],
  query: string,
  limit = MAX_SEARCH_MATCHES,
): { files: string[]; matches: number; truncated: boolean } {
  const needle = query.trim().toLowerCase();
  if (!needle) return { files: [...paths], matches: paths.length, truncated: false };
  const files: string[] = [];
  let matches = 0;
  for (const path of paths) {
    if (!path.toLowerCase().includes(needle)) continue;
    matches += 1;
    if (files.length < limit) files.push(path);
  }
  return { files, matches, truncated: matches > files.length };
}

export type FileTreeRow = { node: FileTreeNode; depth: number };

/**
 * The tree as a flat list of visible rows.
 *
 * FLAT BECAUSE THE KEYBOARD IS FLAT. Up and down move between adjacent VISIBLE
 * rows regardless of nesting, which is what every tree in every file manager
 * does, and expressing that against a nested render is a walk on every keypress.
 * It also means the row count is knowable, which is how the surface can say
 * "showing 400 of 1,127".
 */
export function flattenTree(nodes: readonly FileTreeNode[], expanded: ReadonlySet<string>, depth = 0): FileTreeRow[] {
  const rows: FileTreeRow[] = [];
  for (const node of nodes) {
    rows.push({ node, depth });
    if (node.kind === "directory" && expanded.has(node.path)) rows.push(...flattenTree(node.children, expanded, depth + 1));
  }
  return rows;
}

/** Every directory in the tree — what "expand everything" means while a search
 *  is running, since a filtered tree is small and hiding its matches behind
 *  chevrons would defeat the search. */
export function directoryPaths(nodes: readonly FileTreeNode[]): string[] {
  const paths: string[] = [];
  for (const node of nodes) {
    if (node.kind !== "directory") continue;
    paths.push(node.path);
    paths.push(...directoryPaths(node.children));
  }
  return paths;
}

/**
 * Every directory that contains a changed file, at any depth.
 *
 * So a collapsed `apps/` can carry the mark that something inside it is dirty —
 * which is the whole reason to tint a tree in a tool where an agent is the one
 * doing the writing. Derived from the changed paths rather than by walking the
 * tree, because the changed set is the small one.
 */
export function ancestorsOf(paths: Iterable<string>): Set<string> {
  const dirs = new Set<string>();
  for (const path of paths) {
    const segments = path.split("/");
    segments.pop();
    let prefix = "";
    for (const segment of segments) {
      prefix = prefix ? `${prefix}/${segment}` : segment;
      dirs.add(prefix);
    }
  }
  return dirs;
}

File: apps/ios/TelarMobile/Views/Panel/FileTree.swift
import Foundation

/// The checkout as a tree, built on the phone from the engine's FLAT path
/// list — `apps/web/lib/file-tree.ts`, ported rule for rule.
///
/// Directories before files, then natural order (`step-2` before `step-10`,
/// and `README.md` beside `package.json` rather than above every lowercase
/// name). Single-child directory chains collapse into one row, so a monorepo's
/// `src/main/java/com` reads as one line rather than four.
indirect enum FileTreeNode: Equatable, Identifiable {
    case file(path: String, name: String)
    case directory(path: String, name: String, children: [FileTreeNode])

    var id: String { path }
    var path: String {
        switch self {
        case .file(let path, _), .directory(let path, _, _): path
        }
    }
    var name: String {
        switch self {
        case .file(_, let name), .directory(_, let name, _): name
        }
    }
    var isDirectory: Bool {
        if case .directory = self { return true }
        return false
    }
    var children: [FileTreeNode] {
        if case .directory(_, _, let children) = self { return children }
        return []
    }
}

private final class Building {
    var dirs: [String: Building] = [:]
    var files: [String] = []
}

func buildFileTree(_ paths: [String]) -> [FileTreeNode] {
    let root = Building()
    for path in paths {
        var segments = path.split(separator: "/").map(String.init).filter { !$0.isEmpty }
        guard let name = segments.popLast() else { continue }
        var cursor = root
        for segment in segments {
            if let next = cursor.dirs[segment] {
                cursor = next
            } else {
                let next = Building()
                cursor.dirs[segment] = next
                cursor = next
            }
        }
        cursor.files.append(name)
    }
    return toNodes(root, prefix: "").map(collapse)
}

private func toNodes(_ building: Building, prefix: String) -> [FileTreeNode] {
    var nodes: [FileTreeNode] = []
    for (name, child) in building.dirs {
        let path = prefix.isEmpty ? name : prefix + "/" + name
        nodes.append(.directory(path: path, name: name, children: toNodes(child, prefix: path)))
    }
    for name in building.files {
        nodes.append(.file(path: prefix.isEmpty ? name : prefix + "/" + name, name: name))
    }
    return nodes.sorted(by: compareNodes)
}

private func compareNodes(_ left: FileTreeNode, _ right: FileTreeNode) -> Bool {
    if left.isDirectory != right.isDirectory { return left.isDirectory }
    let order = left.name.compare(right.name, options: [.numeric, .caseInsensitive, .diacriticInsensitive])
    if order == .orderedSame { return left.name < right.name }
    return order == .orderedAscending
}

/// A directory whose only child is a directory folds into it: `a/b/c`.
private func collapse(_ node: FileTreeNode) -> FileTreeNode {
    guard case .directory(let path, let name, let children) = node else { return node }
    let collapsed = children.map(collapse)
    if collapsed.count == 1, case .directory(let childPath, let childName, let grand) = collapsed[0] {
        return .directory(path: childPath, name: name + "/" + childName, children: grand)
    }
    return .directory(path: path, name: name, children: collapsed)
}

struct FileTreeRow: Identifiable, Equatable {
    var node: FileTreeNode
    var depth: Int
    var id: String { node.path }
}

/// The rows a list draws: open directories recurse, closed ones do not.
func flattenTree(_ nodes: [FileTreeNode], expanded: Set<String>, depth: Int = 0) -> [FileTreeRow] {
    var rows: [FileTreeRow] = []
    for node in nodes {
        rows.append(FileTreeRow(node: node, depth: depth))
        if node.isDirectory, expanded.contains(node.path) {
            rows.append(contentsOf: flattenTree(node.children, expanded: expanded, depth: depth + 1))
        }
    }
    return rows
}

/// How many matches a search draws. Typing one letter matches most of a
/// repository; the cap is reported so the surface can say what it dropped.
let maxSearchMatches = 400

/// The paths a query keeps — matched on the WHOLE path, case-insensitively,
/// as a plain substring: a person typing into a file search is remembering a
/// fragment, not writing a pattern.
func matchFiles(_ paths: [String], query: String) -> (matches: [String], dropped: Int) {
    let needle = query.trimmingCharacters(in: .whitespaces).lowercased()
    guard !needle.isEmpty else { return (paths, 0) }
    var matches: [String] = []
    var dropped = 0
    for path in paths where path.lowercased().contains(needle) {
        if matches.count < maxSearchMatches { matches.append(path) } else { dropped += 1 }
    }
    return (matches, dropped)
}

/// Every directory on the way to each path — what a search expands.
func ancestorsOf(_ paths: [String]) -> Set<String> {
    var result = Set<String>()
    for path in paths {
        var segments = path.split(separator: "/").map(String.init)
        segments.removeLast()
        var prefix = ""
        for segment in segments {
            prefix = prefix.isEmpty ? segment : prefix + "/" + segment
            result.insert(prefix)
        }
    }
    return result
}

/// Every directory in the tree, for "expand all" and for mapping a collapsed
/// chain's ancestors back to the row that shows them.
func directoryPaths(_ nodes: [FileTreeNode]) -> [String] {
    var out: [String] = []
    for node in nodes where node.isDirectory {
        out.append(node.path)
        out.append(contentsOf: directoryPaths(node.children))
    }
    return out
}
